### PR TITLE
Add MTP speculative-decoding support to NeMo SpeechLM vLLM plugin

### PR DESCRIPTION
**Pull request by Slyne**
- register the NeMo SpeechLM MTP draft model and route compatible repeated-layer checkpoints through vLLM speculative decoding
- fuse audio embeddings at placeholder positions for target and draft models
- load MTP weights while handling SpeechLM checkpoint prefixes and vocabulary padding
- make plugin registration idempotent and validate repeated-layer multi-head configurations
- add focused plugin tests for registration, config routing, embeddings, and MTP properties

> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?
Add MTP speculative decoding support for NeMo SpeechLM models served through vLLM.

**Collection**:  SpeechLM (`nemo.collections.speechlm2`)

# Changelog

- Register a NeMo SpeechLM MTP draft-model architecture with vLLM.
- Route SpeechLM checkpoints with MTP configuration through vLLM speculative decoding.
- Support repeated-layer, multi-head MTP checkpoints.
- Fuse audio embeddings into placeholder positions for both target and draft models.
- Load MTP checkpoint weights while handling SpeechLM prefixes and vocabulary padding.
- Reject incompatible multi-head checkpoints that were not trained with a repeated MTP layer.
- Make vLLM plugin registration idempotent.
- Add unit tests for registration, configuration routing, weight handling, and multimodal MTP embeddings.

# Usage

- You can potentially add a usage example below

```python
from nemo.collections.speechlm2.vllm.salm import register

register()
```

And then start vllm server:
```
vllm serve /path/to/speechlm-checkpoint \
  --trust-remote-code \
  --speculative-config '{"method":"mtp","num_speculative_tokens":4}'
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- 55 focused SpeechLM vLLM plugin tests.
- vLLM 0.23.0.
- A four-head repeated-layer MTP checkpoint.
- The complete 83,212-sample Open-ASR leaderboard.
- MTP and no-MTP WER remained equivalent within normal run-to-run variation.

### CODE REVIEW

File: nemo/collections/speechlm2/vllm/salm/__init__.py [hunk 1]
```diff
@@ -26,6 +26,84 @@
...
+
+    3. ``ModelRegistry`` — ``NeMoSpeechLMMTPModel`` is registered so that
+       vLLM can resolve and instantiate it as the draft model.
+    """
+    from typing import Literal, get_args
+
+    import vllm.config.speculative as _spec_mod
+    from vllm.config.speculative import SpeculativeConfig
+
+    # 1 ── Extend MTPModelTypes -------------------------------------------
```
Comment (DongjiGao) [inline nemo/collections/speechlm2/vllm/salm/__init__.py]: Can you replace the dash here with a short descriptive comment?
e.g., `# Extend vLLM's recognized MTP model types.`

File: nemo/collections/speechlm2/vllm/salm/__init__.py [hunk 2]
```diff
@@ -26,6 +26,84 @@
...
+        _spec_mod.MTPModelTypes = Literal[old_args + ("nemo_speechlm_mtp",)]
+
+    # 2 ── Wrap hf_config_override ----------------------------------------
+    current_override = SpeculativeConfig.hf_config_override
+    if not getattr(current_override, "_nemo_speechlm_mtp_override", False):
+        original_override = current_override
+
+        def _patched_override(hf_config):
+            if hf_config.model_type == "nemo_speechlm":
+                mtp_cfg = getattr(hf_config, "mtp", {}) or {}
```
Comment (DongjiGao) [inline nemo/collections/speechlm2/vllm/salm/__init__.py]: looks redundant `or {}`

File: nemo/collections/speechlm2/vllm/salm/config.py [hunk 3]
```diff
@@ -178,6 +178,12 @@ def __init__(
 
         self.text_config.vocab_size += _SPEECHLM_EMBED_EXTRA_ROWS
 
+        # vLLM's MTP llm_base_proposer reads image_token_index from the target
+        # model's config to locate multimodal placeholder positions during
+        # speculative decoding. For SpeechLM the <|audio|> token is the first
+        # extra row added above the base backbone vocab.
+        self.image_token_index = self.text_config.vocab_size - _SPEECHLM_EMBED_EXTRA_ROWS
```
Comment (DongjiGao) [inline nemo/collections/speechlm2/vllm/salm/config.py]: nit:
```python
self.image_token_index = self.text_config.vocab_size

self.text_config.vocab_size += _SPEECHLM_EMBED_EXTRA_ROWS
```

File: nemo/collections/speechlm2/vllm/salm/model.py [hunk 4]
```diff
@@ -180,6 +180,33 @@ def embed_multimodal(self, **kwargs) -> MultiModalEmbeddings:
...
+        if multimodal_embeddings is None or is_multimodal is None or not is_multimodal.any():
+            return inputs_embeds
+
+        # Concatenate per-audio embedding tensors and overwrite the audio
+        # placeholder token positions with the actual audio embeddings.
+        audio_embeds = torch.cat(list(multimodal_embeddings), dim=0)
+        inputs_embeds = inputs_embeds.clone()
+        inputs_embeds[is_multimodal] = audio_embeds.to(inputs_embeds.dtype)
+        return inputs_embeds
+
```
Comment (DongjiGao) [inline nemo/collections/speechlm2/vllm/salm/model.py]: Is the target-model override necessary?
`NeMoSpeechLMForConditionalGeneration` already inherits the default implementation from `SupportsMultiModal`

Comment (Slyne) [reply nemo/collections/speechlm2/vllm/salm/model.py]: Thanks. Removed.

File: examples/speechlm2/to_hf.py [hunk 5]
```diff
@@ -102,6 +102,37 @@ def _canonical_torch_dtype_name(dtype: str | torch.dtype) -> str:
...
+                        "A repeated MTP head must serialize exactly one physical layer, but the built LLM "
+                        f"declares num_nextn_predict_layers={actual_mtp_depth}."
+                    )
+            else:
+                # For a preserved native head, the recipe depth is advisory.
+                # Export the physical/logical depth that is actually present.
+                mtp_cfg["num_nextn_predict_layers"] = actual_mtp_depth
     dtype_name = _canonical_torch_dtype_name(dtype)
     config["dtype"] = dtype_name
     config["torch_dtype"] = dtype_name
```
Comment (DongjiGao) [inline examples/speechlm2/to_hf.py:138]: nit: move this part to front, and do early return at line 108 to simplify the logic at line 111
`if not isinstance(mtp_cfg, dict) or not mtp_cfg.get("enabled", False):
    return config`

File: examples/speechlm2/to_hf.py [hunk 6]
```diff
@@ -102,6 +102,37 @@ def _canonical_torch_dtype_name(dtype: str | torch.dtype) -> str:
...
+            if isinstance(actual_mtp_depth, bool) or not isinstance(actual_mtp_depth, int) or actual_mtp_depth <= 0:
+                raise ValueError(
+                    f"Built LLM has invalid num_nextn_predict_layers={actual_mtp_depth!r}; cannot export it."
+                )
+            if mtp_cfg.get("use_repeated_layer", False):
+                if actual_mtp_depth != 1:
+                    raise ValueError(
+                        "A repeated MTP head must serialize exactly one physical layer, but the built LLM "
+                        f"declares num_nextn_predict_layers={actual_mtp_depth}."
+                    )
```
Comment (DongjiGao) [inline examples/speechlm2/to_hf.py:131]: if those check need frequent updates with MTP arch updates, consider centralize them:

```
def _read_mtp_depth(llm_config: Any) -> int | None:
    depth = getattr(
        llm_config,
        "num_nextn_predict_layers",
        None,
    )
   ...

    return depth
```

**Review by DongjiGao — APPROVED** [review]
LGTM. Just 2 minor comments.

**Comment by pzelasko** [comment]
I'm closing this PR in favor of https://github.com/NVIDIA-NeMo/Speech/pull/16158 (same code but re-wired through stack of 6 PRs)
